File: bindings/typescript/src/group.ts
import { ElementModPHandle, ElementModQHandle, getInstance } from "./wasm";

export class ElementModP {
  _handle: ElementModPHandle;

  constructor(handle: ElementModPHandle) {
    this._handle = handle;
  }

  copy(): Promise<ElementModP> {
    var hex = this._handle.toHex();
    return ElementModP.fromHex(hex);
  }

  toHex(): string {
    return this._handle.toHex();
  }

  static async fromNumber(a: number): Promise<ElementModP> {
    var result = await this.fromBigInt(BigInt(a));
    return result;
  }
  static async fromBigInt(a: BigInt): Promise<ElementModP> {
    var result = (await getInstance()).ElementModP.fromUint64(a, false);
    return new ElementModP(result);
  }
  static async fromHex(a: string): Promise<ElementModP> {
    var result = (await getInstance()).ElementModP.fromHex(a, false);
    return new ElementModP(result);
  }
}

export class ElementModQ {
  _handle: ElementModQHandle;

  constructor(handle: ElementModQHandle) {
    this._handle = handle;
  }

  copy(): Promise<ElementModQ> {
    var hex = this._handle.toHex();
    return ElementModQ.fromHex(hex);
  }

  toHex(): string {
    return this._handle.toHex();
  }

  static async fromNumber(a: number): Promise<ElementModQ> {
    var result = await this.fromBigInt(BigInt(a));
    return result;
  }
  static async fromBigInt(a: BigInt): Promise<ElementModQ> {
    var result = (await getInstance()).ElementModQ.fromUint64(a, false);
    return new ElementModQ(result);
  }
  static async fromHex(a: string): Promise<ElementModQ> {
    var result = (await getInstance()).ElementModQ.fromHex(a, false);
    return new ElementModQ(result);
  }
}

export class GroupFunctions {
  static async addModQ(a: number, b: number): Promise<BigInt> {
    var elem1 = await ElementModQ.fromNumber(a);
    var elem2 = await ElementModQ.fromNumber(b);
    var result = (await getInstance()).GroupFunctions.addModQ(
      elem1._handle,
      elem2._handle
    );
    return BigInt("0x" + result.toHex());
  }

  static async randomElementModP(): Promise<ElementModP> {
    var result = (await getInstance()).GroupFunctions.randomElementModP();
    return new ElementModP(result);
  }

  static async randomElementModQ(): Promise<ElementModQ> {
    var result = (await getInstance()).GroupFunctions.randomElementModQ();
    return new ElementModQ(result);
  }
}

File: src/electionguard/wasm/group.cpp
#include "electionguard/group.hpp"

#include "../log.hpp"

#include <emscripten/bind.h>
#include <emscripten/html5.h>
#include <iostream>

using namespace emscripten;
using namespace electionguard;
using namespace std;

class GroupFunctions
{
  public:
    static std::unique_ptr<ElementModQ> addModQ(ElementModQ &a, ElementModQ &b)
    {
        //std::cout << "addModQ " << a.toHex() << std::endl;
        auto result = add_mod_q(a, b);
        return std::move(result);
    }

    static std::unique_ptr<ElementModP> randomElementModP() { return rand_p(); }

    static std::unique_ptr<ElementModQ> randomElementModQ() { return rand_q(); }
};

EMSCRIPTEN_BINDINGS(electionguard)
{
    class_<ElementModP>("ElementModP")
      .function("isInBounds", &ElementModP::isInBounds)
      .function("toHex", &ElementModP::toHex)
      .function("copy", &ElementModQ::clone)
      .class_function("fromHex", &ElementModP::fromHex)
      .class_function("fromUint64", &ElementModP::fromUint64);

    class_<ElementModQ>("ElementModQ")
      .function("isInBounds", &ElementModQ::isInBounds)
      .function("toHex", &ElementModQ::toHex)
      .function("copy", &ElementModQ::clone)
      .function("toElementModP", &ElementModQ::toElementModP)
      .class_function("fromHex", &ElementModQ::fromHex)
      .class_function("fromUint64", &ElementModQ::fromUint64);

    class_<GroupFunctions>("GroupFunctions")
      .class_function("addModQ", &GroupFunctions::addModQ)
      .class_function("randomElementModP", &GroupFunctions::randomElementModP)
      .class_function("randomElementModQ", &GroupFunctions::randomElementModQ);
}